File: web/src/main/java/hu/bsstudio/robonaut/speed/JuniorSpeedRaceResultHandler.java
package hu.bsstudio.robonaut.speed;

import hu.bsstudio.robonaut.race.speed.SpeedRaceService;
import hu.bsstudio.robonaut.race.speed.model.SpeedRaceResult;
import hu.bsstudio.robonaut.team.model.DetailedTeam;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;
import org.springframework.web.reactive.function.server.HandlerFunction;
import org.springframework.web.reactive.function.server.ServerRequest;
import org.springframework.web.reactive.function.server.ServerResponse;
import reactor.core.publisher.Mono;

@RequiredArgsConstructor
public class JuniorSpeedRaceResultHandler implements HandlerFunction<ServerResponse> {

    @NonNull
    private final SpeedRaceService service;

    @Override
    public Mono<ServerResponse> handle(final ServerRequest request) {
        final var detailedTeam = request.bodyToMono(SpeedRaceResult.class)
            .flatMap(service::updateSpeedRaceJunior);
        return ServerResponse.ok().body(detailedTeam, DetailedTeam.class);
    }
}

File: web/src/main/kotlin/hu/bsstudio/robonaut/speed/JuniorSpeedRaceResultHandler.kt
package hu.bsstudio.robonaut.speed

import hu.bsstudio.robonaut.race.speed.SpeedRaceService
import hu.bsstudio.robonaut.race.speed.model.SpeedRaceResult
import org.springframework.web.reactive.function.server.*
import reactor.core.publisher.Mono

class JuniorSpeedRaceResultHandler(private val service: SpeedRaceService) : HandlerFunction<ServerResponse> {

    override fun handle(request: ServerRequest): Mono<ServerResponse> {
        return request.bodyToMono<SpeedRaceResult>()
            .flatMap(service::updateSpeedRaceJunior)
            .let { detailedTeam -> ServerResponse.ok().body(detailedTeam) }
    }
}
